File: blue-core/src/blue/noteProcessor/RandomAddProcessor.java
package blue.noteProcessor;

import blue.BlueSystem;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;
import electric.xml.Elements;
import java.io.Serializable;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

public class RandomAddProcessor implements NoteProcessor, Serializable {

    int pfield = 4;

    float min = 0f;

    float max = 1f;

    public RandomAddProcessor() {
    }

    @Override
    public String toString() {
        return "[random add]";
    }

    public String getPfield() {
        return Integer.toString(pfield);
    }

    public void setPfield(String pfield) {
        this.pfield = Integer.parseInt(pfield);
    }

    public String getMin() {
        return Float.toString(min);
    }

    public void setMin(String value) {
        this.min = Float.parseFloat(value);
    }

    public String getMax() {
        return Float.toString(max);
    }

    public void setMax(String value) {
        this.max = Float.parseFloat(value);
    }

    public final void processNotes(NoteList in) throws NoteProcessorException {
        Note temp;

        float range = max - min;
        float fieldVal = 0f;

        for (int i = 0; i < in.size(); i++) {
            temp = in.getNote(i);
            try {
                fieldVal = Float.parseFloat(temp.getPField(pfield));
            } catch (NumberFormatException ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.pfieldNotFloat"),
                        pfield);
            } catch (Exception ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.missingPfield"),
                        pfield);
            }
            float randVal = (float) ((Math.random() * range) + min);

            temp.setPField(Float.toString(fieldVal + randVal), pfield);
        }
    }

    public static NoteProcessor loadFromXML(Element data) {
        RandomAddProcessor rap = new RandomAddProcessor();

        Elements nodes = data.getElements();

        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            switch (node.getName()) {
                case "pfield":
                    rap.setPfield(node.getTextString());
                    break;
                case "min":
                    rap.setMin(node.getTextString());
                    break;
                case "max":
                    rap.setMax(node.getTextString());
                    break;
            }
        }

        return rap;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("pfield").setText(this.getPfield());
        retVal.addElement("min").setText(this.getMin());
        retVal.addElement("max").setText(this.getMax());

        return retVal;
    }

    public static void main(String args[]) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.addNote(Note.createNote("i1 " + (i * 2) + " 2 3 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        RandomAddProcessor rap = new RandomAddProcessor();
        rap.setPfield("4");
        rap.setMin("0f");
        rap.setMax("1.5f");
        try {
            rap.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");
    }
}

File: packages/blue-data/src/note-processors/random-add-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';
import { JavaRandom } from '../sound-objects/jmask-support';

const JAVA_TYPE = 'blue.noteProcessor.RandomAddProcessor';

export class RandomAddProcessor extends NoteProcessor {
  private _pfield = 4;
  private _min = 0.0;
  private _max = 1.0;
  private _seedUsed = false;
  private _seed = 0;

  constructor();
  constructor(src: RandomAddProcessor);
  constructor(src?: RandomAddProcessor) {
    super();
    if (src) {
      this._pfield = src._pfield;
      this._min = src._min;
      this._max = src._max;
      this._seedUsed = src._seedUsed;
      this._seed = src._seed;
    }
  }

  getPfield(): string {
    return this._pfield.toString();
  }
  setPfield(pfield: string): void {
    this._pfield = parseInt(pfield, 10);
  }

  getMin(): string {
    return this._min.toString();
  }
  setMin(value: string): void {
    this._min = parseFloat(value);
  }

  getMax(): string {
    return this._max.toString();
  }
  setMax(value: string): void {
    this._max = parseFloat(value);
  }

  isSeedUsed(): boolean {
    return this._seedUsed;
  }
  setSeedUsed(seedUsed: boolean): void {
    this._seedUsed = seedUsed;
  }

  getSeed(): string {
    return this._seed.toString();
  }
  setSeed(seed: string): void {
    this._seed = parseInt(seed, 10);
  }

  override process(notes: NoteList): NoteList {
    const range = this._max - this._min;
    const r = this._seedUsed ? new JavaRandom(this._seed) : null;

    for (const note of notes) {
      let fieldVal: number;
      try {
        fieldVal = parseFloat(note.getPField(this._pfield)!);
      } catch {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }
      if (isNaN(fieldVal)) {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }
      const randVal = (r ? r.nextDouble() : Math.random()) * range + this._min;
      note.setPField((fieldVal + randVal).toString(), this._pfield);
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'RandomAddProcessor';
  }

  override deepCopy(): RandomAddProcessor {
    return new RandomAddProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('pfield').setText(this.getPfield());
    elem.addElement('min').setText(this.getMin());
    elem.addElement('max').setText(this.getMax());
    elem.addElement('seedUsed').setText(this._seedUsed.toString());
    elem.addElement('seed').setText(this.getSeed());
    return elem;
  }

  static loadFromXML(data: Element): RandomAddProcessor {
    const proc = new RandomAddProcessor();
    const pf = data.getTextString('pfield');
    if (pf !== null) proc._pfield = parseInt(pf, 10);
    const mn = data.getTextString('min');
    if (mn !== null) proc._min = parseFloat(mn);
    const mx = data.getTextString('max');
    if (mx !== null) proc._max = parseFloat(mx);
    const su = data.getTextString('seedUsed');
    if (su !== null) proc._seedUsed = su.toLowerCase() === 'true';
    const sd = data.getTextString('seed');
    if (sd !== null) proc._seed = parseInt(sd, 10);
    return proc;
  }
}
